lib/active_record/events/method_factory.rb

```diff
@@ -48,7 +48,11 @@ def define_inverse_predicate_method(module_, naming)
 
       def define_action_method(module_, naming)
         module_.send(:define_method, naming.action) do
-          touch(naming.field)
+          if persisted?
+            touch(naming.field)
+          else
+            send("#{naming.field}=", Time.zone.now)
```

Comment:
We could set the attribute using the [`[]=` syntax](https://api.rubyonrails.org/classes/ActiveRecord/AttributeMethods.html#method-i-5B-5D-3D):

```ruby
self[naming.field] = Time.current
```